warehouse/query-core/src/main/java/datawave/query/language/parser/jexl/JexlNodeSet.java

```diff
@@ -0,0 +1,185 @@
+    public boolean isEmpty() {
+        return nodeMap.isEmpty();
+    }
+    
+    @Override
+    public boolean contains(Object o) {
+        if (o instanceof JexlNode) {
+            String nodeKey = JexlASTHelper.nodeToKey((JexlNode) o);
+            return nodeMap.containsKey(nodeKey);
+        }
+        return false;
+    }
+    
+    @Override
+    public Iterator<JexlNode> iterator() {
+        return Lists.newArrayList(nodeMap.values()).listIterator();
+    }
+    
+    @Override
+    public Object[] toArray() {
+        return nodeMap.entrySet().toArray();
+    }
+    
+    @Override
+    public <T> T[] toArray(T[] ts) {
+        throw new UnsupportedOperationException("JexlNodeSet does not support toArray() operations.");
+    }
+    
+    @Override
+    public boolean add(JexlNode node) {
+        String nodeKey = JexlASTHelper.nodeToKey(node);
+        if (!nodeMap.containsKey(nodeKey)) {
+            nodeMap.put(nodeKey, node);
+            return true;
+        }
+        return false;
+    }
+    
+    /**
+     *
+     * @param o
+     *            - object to be removed from the set of Jexl nodes.
+     * @return - True if the set contained the specified element.
+     */
+    @Override
+    public boolean remove(Object o) {
+        if (o instanceof JexlNode) {
+            String nodeKey = JexlASTHelper.nodeToKey((JexlNode) o);
+            return remove(nodeKey, o);
+        }
+        return false;
+    }
+    
+    /**
+     *
+     * @param nodeKey
+     *            - key for the object to be removed.
+     * @param o
+     *            - the object to be removed.
+     * @return - True, if the set contained the specified element.
+     */
+    private boolean remove(String nodeKey, Object o) {
+        return nodeMap.remove(nodeKey, o);
+    }
+    
+    @Override
+    public boolean containsAll(Collection<?> collection) {
+        if (collection != null) {
+            for (Object o : collection) {
+                if (!contains(o)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+        return false;
+    }
+    
+    @Override
+    public boolean addAll(Collection<? extends JexlNode> collection) {
+        boolean modified = false;
+        if (collection != null) {
+            for (Object o : collection) {
+                if (o instanceof JexlNode) {
+                    if (add((JexlNode) o)) {
+                        modified = true;
+                    }
+                }
+            }
+        }
+        return modified;
+    }
+    
+    @Override
+    public boolean retainAll(Collection<?> collection) {
+        Set<String> retainKeys = new HashSet<>();
+        if (collection != null) {
+            for (Object o : collection) {
+                retainKeys.add(JexlASTHelper.nodeToKey((JexlNode) o));
+            }
+        }
+        
+        boolean modified = false;
+        for (String key : Sets.newHashSet(nodeMap.keySet())) {
```

Comment:
Seems unnecessary to create an extra HashSet here if we are only iterating over it.

Reply:
Nope, I'll remove that.